Sprint 12

dkmenshikov opened the pull request.

Commit:
Список изменений:
1) реализован сервис загрузки информации о фото

Commit:
Список изменений:
1) в сервис ImagesListPhotos добавлена обработка случая повторного фетча до завершения первого
2) в сервис ImagesListPhotos добавлена нотификация после обновления массива photos
3) в сервис ImagesListPhotos добавлено заполнение массива photos из response

Commit:
Список изменений:
1) ImagesListPhotos переименован в ImagesListService

Commit:
Список изменений:
1) реализована загрузка ленты фотографиий и подстановка превью в таблицу

Commit:
Список изменений:
1) загрузка следующей страницы с фотографиями осуществляется в момент за 2 фото до конца текущей загруженной ленты
2) исправлено дублирование фотографий в страницах

Commit:
Список изменений:
1) добавлен функционал лайков

Commit:
Список изменений:
1) проект подготовлен для установки картинки на экран single image vc

Commit:
Список изменений:
1) добавлен функционал показа изображения на экране SingleImageViewController

Commit:
Список изменений:
1) добавлен функционал логаута из профиля

Commit:
Список изменений:
1) Добавлена обработка ситуации отсутствия даты фотографии в ответе с сервера
2) исправлен цвет ProgressHUD

Commit:
Список изменений:
1) рационализировано использование DateFormatter в ImagesListService
2) изменено логирование NetworkClient для отображения ошибки и сырого вида JSON в случае невозможности декодирования

Commit:
Список изменений:
1) исправлен force unwrap для indexPath в методе likeButtonDidTap

Commit:
Список изменений:
1) исправлен DateFormatter на ISO8601
2) переменные сервисов в некоторых классах исправлены на константы

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImagesListCell.swift:
```diff
@@ -25,21 +28,36 @@ final class ImagesListCell: UITableViewCell {
         }
+        likeButton.setImage(photo.isLiked ? .likeActive : .likeInactive, for: [])
+        self.delegate = delegate
+        self.indexPath = indexPath
```
Немного странная история с self.indexPath, конечно
В идеале не нужно вручную хранить indexPath, эта вещь работает под капотом через делегат UITableView
Всё можно упростить создав метод setLike внутри ячейки, и не жонглировать делегатами и indexPath)

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImagesListCell.swift:
```diff
@@ -25,21 +28,36 @@ final class ImagesListCell: UITableViewCell {
+        cellPicture.kf.cancelDownloadTask()
     }
     
+    @IBAction func likeButtonDidTap(_ sender: Any) {
```
IBAction тоже стоит делать приватными)

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -6,24 +6,68 @@
     
 //    MARK: - IBOutlets
     
     @IBOutlet weak var tableView: UITableView!
```
И IBOutlet тоже можно делать private

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -6,24 +6,68 @@
+                completion(true)
+            case .failure(let error):
+                print("[LOG]: failure of updating like status, \(error)")
+                UIBlockingProgressHUD.dismiss()
```
Дублирование UIBlockingProgressHUD. Его можно вывести за пределы switch, так как его вызывать в любом случае)

andrew-kuzn-yndx commented on ImageFeed/ImagesList/ImagesListViewController.swift:
```diff
@@ -6,24 +6,68 @@
         super.viewDidLoad()
         setTableView()
+        imagesListServiceObserver = NotificationCenter.default
+            .addObserver(
```
Сейчас не принципиально, но в будущем всегда учитывай, что при добавлении Observer, нужно предусматривать и его удаление, чтобы не создать утечку памяти. Обычно удаление обсервера добавляют в один из заключительных этапов жизненного цикла вьюконтроллера

Commit:
Список изменений:
1) сетевые сервисы объединены в отдельную папку
2) создан сервис AlertPresenter
3) все показы алертов отрефакторены на использование сервиса
4) изменены уровни доступа на private у всех IBActions и IBOutlets
5) добавлен показ алерта при нажатии на кнопку выхода из профиля на экране ProfileViewController
